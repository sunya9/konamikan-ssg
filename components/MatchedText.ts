import Vue, { PropOptions } from 'vue'
import Fuse from 'fuse.js'
import unicodeSubstring from 'unicode-substring'
export default Vue.extend({
  name: 'MatchedText',
  functional: true,
  props: {
    matches: {
      type: Array,
      required: true
    } as PropOptions<Fuse.FuseResultMatch[]>
  },
  render(h, context) {
    const match = context.props.matches[0]
    const [start, originalEnd] = match.indices[0]
    const end = originalEnd + 1
    const { value } = match
    if (!value) return h('span')
    const endOffset = start + 30 < value.length ? end + 30 : value.length
    const suffix = unicodeSubstring(value, end, endOffset)
    const mark = h('mark', unicodeSubstring(value, start, end))
    return h(
      'p',
      {
        staticClass: 'is-text-truncated'
      },
      [mark, suffix]
    )
  }
})
