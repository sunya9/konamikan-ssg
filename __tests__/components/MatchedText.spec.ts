import { mount } from '@vue/test-utils'
import Fuse from 'fuse.js'
import MatchedText from '@/components/MatchedText'

describe('MatchedText', () => {
  it('normal', () => {
    const matches: Fuse.FuseResultMatch[] = [
      {
        key: 'test',
        refIndex: 1,
        value: 'lorem ipsum',
        indices: [[0, 4]]
      }
    ]
    const wrapper = mount(MatchedText, {
      propsData: {
        matches
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.element.tagName).toBe('P')
    expect(wrapper.text()).toBe('lorem ipsum')
    expect(wrapper.find('mark').text()).toBe('lorem')
  })
})
