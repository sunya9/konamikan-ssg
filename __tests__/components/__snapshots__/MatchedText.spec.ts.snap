// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchedText normal 1`] = `<p class="is-text-truncated"><mark>lorem</mark> ipsum</p>`;
